bee-network/src/swarm/protocols/gossip/gossip.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use super::GossipHandler;
+
+use crate::network::meta::{ConnectionInfo, Origin};
+
+use libp2p::{
+    core::{connection::ConnectionId, ConnectedPoint},
+    swarm::{NegotiatedSubstream, NetworkBehaviour, NetworkBehaviourAction, ProtocolsHandler},
+    Multiaddr, PeerId,
+};
+use tokio::sync::mpsc;
+
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::atomic::{AtomicBool, Ordering},
+    task::Poll,
+};
+
+// false == inbound
+pub static GOSSIP_ORIGIN: AtomicBool = AtomicBool::new(false);
+
+#[derive(Eq, PartialEq, Hash, Debug, Clone, Copy)]
+struct Id(PeerId, ConnectionId);
+
+// #[derive(Default)]
+pub struct Gossip {
+    // Connection origins we are listening to
+    // origin_rx: mpsc::UnboundedReceiver<Origin>,
+    // Gossip event builder per peer id
+    builders: HashMap<PeerId, GossipEventBuilder>,
+    // Events produced by the 'GossipHandlers'
+    events: VecDeque<GossipEvent>,
+}
+
+impl Gossip {
+    pub fn new(_origin_rx: mpsc::UnboundedReceiver<Origin>) -> Self {
+        Self {
+            // origin_rx,
+            builders: HashMap::default(),
+            events: VecDeque::default(),
+        }
+    }
+}
+
+impl NetworkBehaviour for Gossip {
+    type ProtocolsHandler = GossipHandler;
+    type OutEvent = GossipEvent; //<Self::ProtocolsHandler as ProtocolsHandler>::OutEvent;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        // NB: it is very hackish how we solve this. There has to be a better way!
+
+        // let origin = if let Some(origin) = self.origin_rx.recv().now_or_never().flatten() {
+        //     origin
+        // } else {
+        //     Origin::Inbound
+        // };
+
+        let origin = if GOSSIP_ORIGIN.swap(false, Ordering::SeqCst) {
+            Origin::Outbound
+        } else {
+            Origin::Inbound
+        };
+
+        // println!("GOSSIP: New_handler: {}", origin);
+        GossipHandler::new(origin)
+    }
+
+    fn addresses_of_peer(&mut self, _peer_id: &libp2p::PeerId) -> Vec<libp2p::Multiaddr> {
+        // println!("Addresses of peer: {}", peer_id);
+        Vec::new()
+    }
+
+    fn inject_connection_established(&mut self, peer_id: &PeerId, conn_id: &ConnectionId, endpoint: &ConnectedPoint) {
+        // TODO: Perform the connection checks (not banned, not a duplicate etc)
```

Comment:
I'll keep it for now. Need to think about this a bit more. I cannot easily get the peerlist in there without resorting to making it global state which I'ld like to prevent.